File: tests/fixtures/python-domain/src/payment/entity.py
"""Payment domain: entity, value objects and invariants."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import Enum


class InvalidTransitionError(Exception):
    """Raised when a status transition is not allowed."""


@dataclass(frozen=True)
class Money:
    """Value object representing a monetary amount with currency."""

    amount: int
    currency: str

    def __post_init__(self) -> None:
        if self.amount < 0:
            raise ValueError("amount must be non-negative")
        if not self.currency:
            raise ValueError("currency is required")


class PaymentStatus(str, Enum):
    PENDING = "pending"
    CONFIRMED = "confirmed"
    CANCELLED = "cancelled"


@dataclass
class Payment:
    """Aggregate root for the payment domain."""

    id: str
    customer_id: str
    amount: Money
    status: PaymentStatus = field(default=PaymentStatus.PENDING)
    created_at: datetime = field(default_factory=lambda: datetime.now(timezone.utc))

    def __post_init__(self) -> None:
        if not self.id:
            raise ValueError("id is required")
        if not self.customer_id:
            raise ValueError("customer_id is required")

    def confirm(self) -> None:
        if self.status != PaymentStatus.PENDING:
            raise InvalidTransitionError(
                f"cannot confirm from {self.status.value}"
            )
        self.status = PaymentStatus.CONFIRMED

    def cancel(self) -> None:
        if self.status == PaymentStatus.CANCELLED:
            raise InvalidTransitionError("already cancelled")
        self.status = PaymentStatus.CANCELLED

File: tests/fixtures/node-domain/src/payment/entity.ts
/** Money is a value object representing a monetary amount with currency. */
export class Money {
  readonly amount: number;
  readonly currency: string;

  private constructor(amount: number, currency: string) {
    this.amount = amount;
    this.currency = currency;
  }

  static create(amount: number, currency: string): Money {
    if (amount < 0) throw new Error("amount must be non-negative");
    if (!currency) throw new Error("currency is required");
    return new Money(amount, currency);
  }
}

export type PaymentStatus = "pending" | "confirmed" | "cancelled";

export class InvalidTransitionError extends Error {
  constructor(from: PaymentStatus, to: string) {
    super(`invalid transition from ${from} to ${to}`);
    this.name = "InvalidTransitionError";
  }
}

/** Payment is the aggregate root for the payment domain. */
export class Payment {
  readonly id: string;
  readonly customerId: string;
  readonly amount: Money;
  private _status: PaymentStatus;
  readonly createdAt: Date;

  private constructor(id: string, customerId: string, amount: Money) {
    this.id = id;
    this.customerId = customerId;
    this.amount = amount;
    this._status = "pending";
    this.createdAt = new Date();
  }

  get status(): PaymentStatus {
    return this._status;
  }

  static create(id: string, customerId: string, amount: Money): Payment {
    if (!id) throw new Error("id is required");
    if (!customerId) throw new Error("customerId is required");
    return new Payment(id, customerId, amount);
  }

  confirm(): void {
    if (this._status !== "pending") {
      throw new InvalidTransitionError(this._status, "confirmed");
    }
    this._status = "confirmed";
  }

  cancel(): void {
    if (this._status === "cancelled") {
      throw new InvalidTransitionError(this._status, "cancelled");
    }
    this._status = "cancelled";
  }
}
